test/pipelinefinally_test.go

```diff
@@ -473,6 +480,246 @@ func TestPipelineLevelFinally_OneFinalTaskFailed_Failure(t *testing.T) {
+	if _, err := c.TaskClient.Create(ctx, finalTask2, metav1.CreateOptions{}); err != nil {
+		t.Fatalf("Failed to create final Task: %s", err)
+	}
+
+	pt := map[string]pipelineTask{
+		"dagtask1": {
+			TaskName: task1.Name,
+		},
+		"dagtask2": {
+			TaskName: task2.Name,
+			Param: []v1beta1.Param{{
+				Name: "dagtask1-result",
+				Value: v1beta1.ArrayOrString{
+					Type:      "string",
+					StringVal: "$(tasks.dagtask1.results.result)",
+				},
+			}},
+		},
+	}
+
+	fpt := map[string]pipelineTask{
+		"finaltask1": {
+			TaskName: finalTask1.Name,
+		},
+		"finaltask2": {
+			TaskName: finalTask2.Name,
+			Param: []v1beta1.Param{{
+				Name: "dagtask1-result",
+				Value: v1beta1.ArrayOrString{
+					Type:      "string",
+					StringVal: "$(tasks.dagtask1.results.result)",
+				},
+			}},
+		},
+	}
+
+	pipeline := getPipeline(t, namespace, pt, fpt)
+	if _, err := c.PipelineClient.Create(ctx, pipeline, metav1.CreateOptions{}); err != nil {
+		t.Fatalf("Failed to create Pipeline: %s", err)
+	}
+
+	pipelineRun := getPipelineRun(t, namespace, pipeline.Name)
+	if _, err := c.PipelineRunClient.Create(ctx, pipelineRun, metav1.CreateOptions{}); err != nil {
+		t.Fatalf("Failed to create Pipeline Run `%s`: %s", pipelineRun.Name, err)
+	}
+
+	if err := WaitForPipelineRunState(ctx, c, pipelineRun.Name, timeout, Running(pipelineRun.Name), "PipelineRunRunning"); err != nil {
+		t.Errorf("Error waiting for PipelineRun %s to start: %s", pipelineRun.Name, err)
+		t.Fatalf("PipelineRun execution failed")
+	}
+
+	patches := []jsonpatch.JsonPatchOperation{{
+		Operation: "add",
+		Path:      "/spec/status",
+		Value:     v1beta1.PipelineRunSpecStatusCancelledRunFinally,
+	}}
+	patchBytes, err := json.Marshal(patches)
+	if err != nil {
+		t.Fatalf("failed to marshal patch bytes in order to stop")
+	}
+	if _, err := c.PipelineRunClient.Patch(ctx, pipelineRun.Name, types.JSONPatchType, patchBytes, metav1.PatchOptions{}, ""); err != nil {
+		t.Fatalf("Failed to patch PipelineRun `%s` with graceful stop: %s", pipelineRun.Name, err)
+	}
+
+	if err := WaitForPipelineRunState(ctx, c, pipelineRun.Name, timeout, FailedWithReason(v1beta1.PipelineRunReasonCancelled.String(), pipelineRun.Name), "PipelineRunCancelled"); err != nil {
+		t.Errorf("Error waiting for PipelineRun %s to finish: %s", pipelineRun.Name, err)
+		t.Fatalf("PipelineRun execution failed")
+	}
+
+	taskrunList, err := c.TaskRunClient.List(ctx, metav1.ListOptions{LabelSelector: "tekton.dev/pipelineRun=" + pipelineRun.Name})
+	if err != nil {
+		t.Fatalf("Error listing TaskRuns for PipelineRun %s: %s", pipelineRun.Name, err)
+	}
+
+	// verify dag task succeeded and final task failed
+	for _, taskrunItem := range taskrunList.Items {
+		switch n := taskrunItem.Labels["tekton.dev/pipelineTask"]; n {
+		case "dagtask1":
+			if !isCancelled(t, n, taskrunItem.Status.Conditions) {
+				t.Fatalf("dag task %s should have been cancelled", n)
+			}
+		case "finaltask1":
```

Comment:
we could add similar check on `finaltask2`, making sure that final task was skipped since its consuming a task result produced by cancelled dag task, please feel free to delay adding this check to a followup PR.